src/flask_pyoidc/factory/cache_factory.py

```diff
@@ -0,0 +1,34 @@
+import time
+
+import cachetools
+from oic.extension.message import TokenIntrospectionResponse
+
+
+class TokenIntrospectionCacheFactory(cachetools.TTLCache):
+    """Time aware caching factory for token introspection parameters and response."""
+    def __setitem__(self, key, token_introspection_response: TokenIntrospectionResponse,
+                    cache_setitem=cachetools.Cache.__setitem__) -> None:
+        """Stores cache of token introspection.
+
+        Parameters
+        ----------
+        key: cachetools.keys._HashedTuple
+            Token introspection parameters.
+        token_introspection_response: TokenIntrospectionResponse
+
+        Notes
+        -----
+        To maintain user security, the cached access token must be purged at
+        the expiry time of access token or at the expiry time of cache itself
+        or whatever happens first. This method sets custom ttl for the cache,
+        based on what happens first.
+        """
+        time_to_live = self.ttl
+        if token_introspection_response.get('active', False):
+            expires_in_seconds = token_introspection_response['exp'] - time.time()
+            if expires_in_seconds < time_to_live:
+                # Set the attribute to the expiry time of access token.
+                self._TTLCache__ttl = expires_in_seconds
```

Comment:
This does not look to be safe - it modifies common state of the cache per request and it does it using a private interface.
Unfortunately I don't think this is a suitable solution to this.